lib/Sema/CSStep.h

```diff
@@ -652,6 +654,29 @@ class DisjunctionStep final : public BindingStep<DisjunctionChoiceProducer> {
     assert(Disjunction->getKind() == ConstraintKind::Disjunction);
     pruneOverloadSet(Disjunction);
     ++cs.solverState->NumDisjunctions;
+
+    // FIXME: This is duplicate (and expensive) work from simplifyAppliedOverloads
+    auto choices = disjunction->getNestedConstraints();
+    auto *typeVar = choices.front()->getFirstType()->getAs<TypeVariableType>();
+    if (!typeVar)
+      return;
+
+    auto result = cs.findConstraintThroughOptionals(
```

Comment:
This is definitely would be great to remove before merging as per your FIXME because it adds one more usage of `gatherConstraints` which is expensive.

Reply:
✅